Integrate cffi-buildtool into CFFI itself (#241)
python-cffi/cffi: doc/source/buildtool.rst

.. _buildtool_docs:

=========================================
Building and Distributing CFFI Extensions
=========================================

.. contents::

CFFI ships a command-line tool, ``gen-cffi-src``, that produces the
same output as :meth:`FFI.emit_c_code`: a ``.c`` source file ready to

> be compiled into a CPython extension module. This tool enables

Reviewer: ```suggestion
be compiled into a CPython c-extension module. This tool enables
```

integrating with any build backend, such as `meson-python
<https://meson-python.readthedocs.io/>`_, `scikit-build-core
<https://scikit-build-core.readthedocs.io/>`_, or similar.

Installing CFFI installs the ``gen-cffi-src`` script; running ``python

> -m cffi.buildtool`` invokes the same command line and behaves

Reviewer: I still find the name buildtool confusing. What does the module actually do? It does not replace mason/setuptools, does it do more than generate c code?

backend`_ that lets you run a helper program during the build can
drive ``gen-cffi-src`` the same way.

The only way to use the buildtool functionality is via this command
line; the implementation inside the ``cffi`` package is private.

The implementation is based on the `cffi-buildtool`_ project by Rose Davidson
(`@inklesspen`_ on GitHub). It is included in CFFI with permission of the original
author.

.. _Python build backend: https://packaging.python.org/en/latest/guides/tool-recommendations/#build-backends-for-extension-modules
.. _cffi-buildtool: https://github.com/inklesspen/cffi-buildtool
.. _@inklesspen: https://github.com/inklesspen

The ``gen-cffi-src`` Command-line Tool
======================================

``gen-cffi-src`` has two subcommands. The first, ``exec-python`` is
most useful if you already have a Python script that sets up an FFI
definition. The second, ``read-sources`` is most useful if you are wrapping
a large API surface and want a more structured way to specify a set of FFI
definitions.

``gen-cffi-src exec-python``
----------------------------

This mode takes a Python script that dynamically defines an FFI interface and

> accompanying C extension source code. The FFI definition script is the same

Reviewer: And does what exactly? What is the output?

    make_ffi = FFI()

In that case, you would pass ``--ffi-var=make_ffi`` to ``gen-cffi-src``:

.. code-block:: console

    $ gen-cffi-src exec-python --ffi-var=make_ffi _squared_build.py _squared.c

.. note::

   CFFI's setuptools integration supports passing ``libraries=``,
   ``library_dirs=``, ``include_dirs=``, and ``extra_compile_args=``
   arguments to :meth:`FFI.set_source`. When using ``gen-cffi-src``,
   these arguments are *ignored*. Link and include settings are the
   build backend's responsibility; for meson-python you would express
   them through the ``dependencies``, ``include_directories``, and
   ``c_args`` arguments of ``py.extension_module()``.

``gen-cffi-src read-sources``
-----------------------------

For larger modules, keeping the FFI definition and any necessary C
source prelude in separate files tends to be easier to work with --
you can configure your editor to treat them as plain C, and write
presubmit tooling that parses the FFI definition directly without
extracting it from a Python script.

Given ``squared.cdef.txt``:

.. code-block:: C

   int square(int n);

and ``squared.csrc.c``:

.. code-block:: C

   #include "square.h"

> you would run the following command to generate a CFFI extension:

Reviewer: ```suggestion
you would run the following command to generate the c code for a CFFI extension:
```

.. _buildtool_docs:

=========================================
Building and Distributing CFFI Extensions
=========================================

.. contents::

> CFFI ships a small subpackage, :mod:`cffi.buildtool`, together with a

Reviewer: ```suggestion
CFFI ships a subpackage, :mod:`cffi.buildtool`, together with a
```

> command-line program, ``gen-cffi-src``. Both produce the same output as

Reviewer: Is the command line program installed with an entry point?

Reviewer 2: Yes; see the change in pyproject.toml.

Author: On further reflection, I think it simplifies things to make the only UI be `python -m cffi.buildtool`, so I've removed the `gen-cffi-src` script and entry point in the latest version of the PR.

Reviewer 3: I think the previous version you had was better. Specifically, the old version worked much better for cross compilation, because the shebang line in the installed entrypoint will correctly find the native Python interpreter that `cffi` is installed for, rather than the cross Python that's being targeted by the build. (analogous example: we use `cython` as a CLI tool)

I'd revert this particular change.

Author: OK, fair enough, let's re-add the entry point script.

Author: Re-added, see latest commit.

:meth:`FFI.emit_c_code`: a ``.c`` source file ready to be compiled into
a CPython extension module. What they add is two convenient front-ends
-- one that executes an existing "build" Python script, and one that

> reads a ``cdef`` and C prelude from two files. This tool enables

Reviewer: I don't understand. Can you rework the "They add" with more words?
```suggestion
a CPython extension module. They add is two convenient front-ends
-- one that executes an existing "build" Python script, and one that
reads a ``cdef`` and C prelude from two files. This tool enables
```

Author: I'm applying your suggestion but I'm going to push a subsequent comment that rewords this.

Author: This paragraph ended up getting dropped completely. I think the point that `cffi.buildtool` is only for generating C sources is a clearer now.

> integrating with and build backend, such as `meson-python

Author: ```suggestion
integrating with any build backend, such as `meson-python
```

<https://meson-python.readthedocs.io/>`_, `scikit-build-core
<https://scikit-build-core.readthedocs.io/>`_, or similar.

The rest of this page uses meson-python in the examples, but any PEP
517 backend that lets you run a helper program during the build can
drive ``gen-cffi-src`` the same way.

> The ``cffi.buildtool`` subpackage was integrated from the

Reviewer: ```suggestion
The ``cffi.buildtool`` subpackage was vendored with permission from the
```

`cffi-buildtool`_ project by Rose Davidson (@inklesspen on GitHub).

.. _cffi-buildtool: https://github.com/inklesspen/cffi-buildtool

Python API for ``cffi.buildtool``
=================================

.. py:module:: cffi.buildtool

.. py:function:: find_ffi_in_python_script(pysrc, filename, ffivar)

   Execute a Python build script and return the :class:`cffi.FFI`

>    object it defines. ``pysrc`` is the text of the script,

Reviewer: Does the build script always compile a c-extension module? Is the lib object also returned somehow?

Author: No, the build script doesn't build an extension module at all. It generates the FFI binding object (a `cffi.api.FFI` instance) which then is used to generate the C extension sources.

I guess calling it a Python build script is confusing. How about Calling it "a script that dynamically defines a set of FFI bindings"?

I'll try to rewrite these docs to make that point clearer.

   ``filename`` is used for diagnostics, and ``ffivar`` is the name
   the script binds to the :class:`FFI` (or to a callable returning
   one -- typical ``ffibuilder`` names are supported). The script is
   executed with ``__name__`` set to ``"gen-cffi-src"`` so a trailing
   ``if __name__ == "__main__": ffibuilder.compile()`` block is
   skipped.

.. py:function:: make_ffi_from_sources(modulename, cdef, csrc)

   Build an :class:`cffi.FFI` from a ``cdef`` string and a C source
   prelude. Equivalent to::

     ffi = FFI()
     ffi.cdef(cdef)
     ffi.set_source(modulename, csrc)

.. py:function:: generate_c_source(ffi)

   Return the C source that :meth:`FFI.emit_c_code` would write for
   the given :class:`cffi.FFI`, as a :class:`str`.

The ``gen-cffi-src`` Command-line Tool
======================================

``gen-cffi-src`` has two subcommands. In both, the final positional
argument is the path to the ``.c`` file to generate.

.. note::

>    When you drive the build from a build backend, the

Reviewer: What does "drive the build" mean in this context?

Author: I mean when CFFI is used without CFFI's built-in distutils integration. Specifically via functionality `recompiler.py`. I'll rephrase this to make that clearer.

   ``libraries=``, ``library_dirs=``, ``include_dirs=``,
   ``extra_compile_args=`` etc. arguments you pass to
   :meth:`FFI.set_source` are *ignored*. Link and include settings are
   the build backend's responsibility; for meson-python you express

>    them through the ``dependencies:`` / ``include_directories:``

Reviewer 3: `extra_compile_args` maps to `c_args`, might want to add that for completeness

   arguments of ``py.extension_module()``.

``gen-cffi-src exec-python``
----------------------------

This mode takes the Python build script you would normally run by
hand -- the one the CFFI docs show under "Main mode of usage" -- and

> generates the ``.c`` source for you. For example, given this

Reviewer: Does it stop after generating the C source, or does it also compile the c-extension module?

Reviewer 2: It stops after generating the C source; it is the job of a c compiler to compile the c extension module. (Build tools such as meson already have support for this.)

Author: I'll try to make this whole set of documentation clearer that there are no compilation steps involved - this is all about generating the C source code for a CFFI extension at build time, without relying on setuptools or CFFI's recompiler and setuptools integration.

      "squared._squared",
      '#include "square.h"',
  )

  if __name__ == "__main__":
      ffibuilder.compile(verbose=True)

you run:

.. code-block:: console

    $ gen-cffi-src exec-python _squared_build.py _squared.c

If the :class:`cffi.FFI` is bound to a name other than ``ffibuilder``,
pass ``--ffi-var``:

.. code-block:: console

    $ gen-cffi-src exec-python --ffi-var=make_ffi _squared_build.py _squared.c

``gen-cffi-src read-sources``
-----------------------------

For larger modules, keeping the ``cdef`` and the C source prelude in
separate files tends to be easier to work with -- your editor
treats them as plain C, and presubmit tooling doesn't have to parse
them out of a string literal.

Given ``squared.cdef.txt``:

.. code-block:: C

   int square(int n);

and ``squared.csrc.c``:

.. code-block:: C

   #include "square.h"

you run:

.. code-block:: console

   $ gen-cffi-src read-sources squared._squared squared.cdef.txt squared.csrc.c _squared.c

The first positional argument is the fully qualified module name that
will be embedded in the generated source (equivalent to the first
argument to :meth:`FFI.set_source`).

A Worked Example Using ``meson-python``

> =======================================

Author: Maybe this should just link to the example projects in the tests to avoid duplication?

Reviewer 2: I think there's value in having it right there in the docs, personally.

          ├── square.h
          └── square.c

``pyproject.toml``:

.. code-block:: toml

  [build-system]
  build-backend = 'mesonpy'
  requires = ['meson-python', 'cffi']

  [project]
  name = 'squared'
  version = '0.1.0'
  requires-python = '>=3.9'
  dependencies = ['cffi']

``meson.build``:

.. code-block:: meson

  project(
      'squared',
      'c',
      version: '0.1.0',
  )

  py = import('python').find_installation(pure: false)

  install_subdir('src/squared', install_dir: py.get_install_dir())

  gen_cffi_src = find_program('gen-cffi-src')

  square_lib = static_library(
      'square',
      'src/csrc/square.c',
      include_directories: include_directories('src/csrc'),
  )
  square_dep = declare_dependency(
      link_with: square_lib,
      include_directories: include_directories('src/csrc'),
  )

  squared_ext_src = custom_target(
      'squared-cffi-src',
      command: [
          gen_cffi_src,
          'exec-python',
          '@INPUT@',
          '@OUTPUT@',
      ],
      output: '_squared.c',
      input: ['src/squared/_squared_build.py'],
  )

  py.extension_module(
      '_squared',
      squared_ext_src,
      subdir: 'squared',
      install: true,

>       dependencies: [square_dep, py.dependency()],

Reviewer 3: Drop `py.dependency` here, it's never needed inside a `py.extension_module` call. It _may_ be needed inside a `static_library` call (as above), if that used CPython C API rather than it being a pure C library (EDIT: but if it were, you'd see it in CI as a build failure).
